---
title: "RRBS - BiSeq rTg4510 - Manhattan & QQ plots DMPs"
author: "Isabel Castanho"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/gpfs/ts0/projects/Research_Project-191406/isabel/RRBS_new/')

# # load packages
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
#
# BiocManager::install("BiSeq")

library(BiSeq)
library(betareg)
library(gplots)
#library(cgwtools) # package that allows to resave objects in R using resave(..., list = character(), file)
library(stringr)
library(tidyverse)
library(qqman)
library(QQperm) # to calculate lambda (inflation)

color_Tg4510_TG <- "#00AEC9"

# Phenotypic data
coldata <- read.csv("/gpfs/ts0/projects/Research_Project-191406/isabel/RRBS_new/Tg4510_coldata_RRBS.csv", row.names=1, stringsAsFactors=FALSE)
coldata$Age_months <- as.numeric(coldata$Age_months)
coldata$Genotype <- as.factor(coldata$Genotype)
coldata$Genotype <- relevel(coldata$Genotype, "WT")
levels(coldata$Genotype)

# ENTORHINAL CORTEX
class(coldata$ECX)

# Remove samples that have NA for pathology
coldata_ECX <- coldata[,c("Genotype", "Age_months", "Histology_no", "ECX")]
coldata_ECX_clean <- na.omit(coldata_ECX)
coldata_pathology <- coldata_ECX_clean

# output directory for pdf 
output <- "/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/scripts/rrbs-ad-mice/Figures"
```
  
# Beta regression model: Methylation = ~ Genotype + Age + Genotype*Age
  
Prepare the data

```{r}
# get data
stats <- read.csv(file = "/gpfs/ts0/projects/Research_Project-191406/isabel/RRBS_new/DMPs/rTg4510/DMPsInteractionModel_stats_table_rTg4510.csv", row.names=1)

# add in columns needed to make manhattan plot
colnames(stats)[1] <- "SNP"
stats$SNP <- gsub(":", "_", stats$SNP)
stats$CHR <- sapply(1:nrow(stats), function(i) str_extract(stats$SNP[i], "chr[^_]*"))
stats$CHR <- substring(stats$CHR, 4)
stats$BP <- sapply(strsplit(rownames(stats), split=':', fixed=TRUE), function(x) (x[2]))

# change chrX to 20, chrY to 21 and M to 22
stats$CHR <- gsub("X", 20, stats$CHR)
stats$CHR <- gsub("Y", 21, stats$CHR)   
stats$CHR <- gsub("M", 22, stats$CHR)
stats$CHR <- as.numeric(stats$CHR)
stats$BP <- as.numeric(stats$BP)
```


## QQ plots
  
```{r, fig.cap = "QQ plot Genotype (p-value)"}
# Genomic inflation factor lambda
lamda <- qchisq(1-median(stats$p.val.Genotype),1)/qchisq(0.5,1)

# plot
qq(stats$p.val.Genotype, main = "rTg4510 Genotype")
mtext(paste("Lamda = ", signif(lamda,3)), side = 3, adj = 1)
```
  
  
```{r, fig.cap = "QQ plot Age (p-value)"}
# Genomic inflation factor lambda
lamda <- qchisq(1-median(stats$p.val.Age),1)/qchisq(0.5,1)

# plot
qq(stats$p.val.Age, main = "rTg4510 Age")
mtext(paste("Lamda = ", signif(lamda,3)), side = 3, adj = 1)
```
  
  
```{r, fig.cap = "QQ plot Genotype*Age (p-value)"}
# Genomic inflation factor lambda
lamda <- qchisq(1-median(stats$p.val.Interaction),1)/qchisq(0.5,1)

# plot
qq(stats$p.val.Interaction, main = "rTg4510 Genotype*Age")
mtext(paste("Lamda = ", signif(lamda,3)), side = 3, adj = 1)
```
  
  
## Manhattan plots

```{r}
# Remove chrY (all samples are females) and chrM
stats_filtered <- stats %>% filter(CHR < 21)

# Set p-value trehsholds
threshold <- 0.05
sig2 <- 0.01
```


blue line: FDR-corrected p-value = 0.05 [-log10(0.05)]

red line: FDR-corrected p-value = 0.01 [-log10(0.01)]


### Genotype
  
```{r, fig.height = 15, fig.width = 15, fig.cap = "Manhattan plot Genotype"}
pdf(paste0(output,"/rTg4510_manhattan_plot.pdf"),width = 10, height = 10)
manhattan(stats_filtered,
	p = "FDR_adj_genotype",
	ylim = c(0,16),
	ylab = "-log10(FDR-adjusted p-value)",
	cex = 1.5, # point size
	cex.axis = 1.2,
	cex.lab = 2,
	col = c("black", color_Tg4510_TG),
	suggestiveline = -log10(threshold),
	genomewideline = -log10(sig2),
	chrlabs = c(1:19, "X"),
	# annotatePval = 1.947822e-34,
	las = 2, # rotate x axis labels so they all fit on the plot
	annotateTop = FALSE,
	main = "Genotype")
dev.off()

```


Zoomed in peaks that were identified as DMRs

```{r}
# Highlight peak in chr2
manhattan(subset(stats_filtered, CHR == 2),
          p = "FDR_adj_genotype",
          suggestiveline = -log10(threshold),
          genomewideline = -log10(sig2),
          main = "Chr 2")
# to highlight sites use argument: highlight = snpsOfInterest
```


```{r}
# Highlight peak in chr5
manhattan(subset(stats_filtered, CHR == 5),
          p = "FDR_adj_genotype",
          suggestiveline = -log10(threshold),
          genomewideline = -log10(sig2),
          main = "Chr 5")
# to highlight sites use argument: highlight = snpsOfInterest
```


```{r}
# Highlight peak in chr18
manhattan(subset(stats_filtered, CHR == 18),
          p = "FDR_adj_genotype",
          suggestiveline = -log10(threshold),
          genomewideline = -log10(sig2),
          main = "Chr 18")
# to highlight sites use argument: highlight = snpsOfInterest
```


### Age
  
```{r, fig.height = 10, fig.width = 15, fig.cap = "Manhattan plot Age"}
manhattan(stats_filtered,
	p = "FDR_adj_age",
	# ylim = c(0,6),
	ylab = "-log10(FDR-adjusted p-value)",
	cex = 2, # point size
	cex.axis = 2.5,
	cex.lab = 2,
	col = rainbow(22),
	suggestiveline = -log10(threshold),
	genomewideline = -log10(sig2),
	chrlabs = c(1:19, "X"),
	# annotatePval = 0.001,
	las = 2, # rotate x axis labels so they all fit on the plot
	annotateTop = FALSE,
	main = "Age")
```
  
  
### Genotype*Age
  
```{r, fig.height = 10, fig.width = 15, fig.cap = "Manhattan plot Genotype*Age"}
manhattan(stats_filtered,
	p = "FDR_adj_interaction",
	# ylim = c(0,8),
	ylab = "-log10(FDR-adjusted p-value)",
	cex = 2, # point size
	cex.axis = 2.5,
	cex.lab = 2,
	col = c("black", color_Tg4510_TG),
	suggestiveline = -log10(threshold),
	genomewideline = -log10(sig2),
	chrlabs = c(1:19, "X"),
	# annotatePval = 0.00001,
	las = 2, # rotate x axis labels so they all fit on the plot
	annotateTop = FALSE,
	main = "Genotype*Age")
```

---

# Beta regression model: Methylation = ~ Pathology

Prepare the data

```{r}
# get data
stats_path <- read.csv(file = "/lustre/projects/Research_Project-191406/isabel/RRBS_new/DMPs/rTg4510/DMPsPathology_stats_table_rTg4510.csv", row.names=1)

test <- read.csv(file = "/lustre/projects/Research_Project-191406/isabel/RRBS_new/DMPs/rTg4510/DMPsPathology_rTg4510_sig_pathology.csv", row.names=1)

# add in columns needed to make manhattan plot
colnames(stats_path)[1] <- "SNP"
stats_path$SNP <- gsub(":", "_", stats_path$SNP)
stats_path$CHR <- sapply(1:nrow(stats_path), function(i) str_extract(stats_path$SNP[i], "chr[^_]*"))
stats_path$CHR <- substring(stats_path$CHR, 4)
stats_path$BP <- sapply(strsplit(rownames(stats_path), split=':', fixed=TRUE), function(x) (x[2]))

# change chrX to 20, chrY to 21 and M to 22
stats_path$CHR <- gsub("X", 20, stats_path$CHR)
stats_path$CHR <- gsub("Y", 21, stats_path$CHR)   
stats_path$CHR <- gsub("M", 22, stats_path$CHR)
stats_path$CHR <- as.numeric(stats_path$CHR)
stats_path$BP <- as.numeric(stats_path$BP)
```


```{r, fig.cap = "QQ plot Tau pathology (p-value)"}
# Genomic inflation factor lambda
lamda <- qchisq(1-median(stats_path$p.val.Pathology),1)/qchisq(0.5,1)

# plot
qq(stats_path$p.val.Pathology, main = "rTg4510 Tau pathology")
mtext(paste("Lamda = ", signif(lamda,3)), side = 3, adj = 1)
```


```{r}
# Remove chrY (all samples are females) and chrM
stats_path_filtered <- stats_path %>% filter(CHR < 21)
```


```{r, fig.height = 10, fig.width = 15, fig.cap = "Manhattan plot Tau Pathology"}
manhattan(stats_path_filtered,
	p = "FDR_adj_pathology",
	ylim = c(0,400),
	ylab = "-log10(FDR-adjusted p-value)",
	cex = 2, # point size
	cex.axis = 2.5,
	cex.lab = 2,
	col = c("black", color_Tg4510_TG),
	suggestiveline = -log10(threshold),
	genomewideline = -log10(sig2),
	chrlabs = c(1:19, "X"),
	# annotatePval = 0.00001,
	las = 2, # rotate x axis labels so they all fit on the plot
	annotateTop = FALSE,
	main = "Tau pathology")
```