---
title: "J20 Array Beta Regression Results"
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, warning=F, message=F, warning=F, message=F}
library(ggplot2)
library(tidyverse)
library(tidyr)
library(qqman)
library(cgwtools)
library(BiSeq)
library(reshape2)
library(gridExtra)
library(kableExtra)
## J20
# Load the unique dataset
load("/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/Normalised_Data_Sesame.rdat")
source("/gpfs/ts0/projects/Research_Project-191406/Aisha/script/rrbs-ad-mice/PlotFunctions.R")
source("/gpfs/ts0/projects/Research_Project-191406/Aisha/script/rrbs-ad-mice/Array/BetaRegressionArray.R")
# Filter for J20 samples
pheno <- QCmetrics[which(QCmetrics$Tissue == "CortexEntorhinalis" & QCmetrics$AD_model == "J20"),]
betas <- Normalised_Sesame_Betas[,colnames(Normalised_Sesame_Betas) %in% pheno$Basename ]
pheno$Genotype <- as.factor(pheno$Genotype)

# Load pathology data to set ages
pathology_pheno <- read.csv("/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/J20_ECX_Pathology.csv", header = T, stringsAsFactors = F)
colnames(pathology_pheno)[colnames(pathology_pheno) == 'X...Sample_ID_ECX'] <- 'Sample_ID_ECX'
pheno$Sample_ID_ECX <- gsub(".*_", "", pheno$ExternalSampleID)

phenowPath <- merge(pheno,pathology_pheno, by = "Sample_ID_ECX")
colnames(phenowPath)[colnames(phenowPath) == 'Genotype.x'] <- 'Genotype'
pheno <- phenowPath

betas <- betas[,order(colnames(betas))]
pheno <- pheno[order(pheno$Basename),]
identical(colnames(betas), pheno$Basename)
pheno$Genotype <- factor(pheno$Genotype, levels = c("WT","TG"))
pheno$Chip_ID <- pheno$Slide

#Load coordinates matrix
probe_mapping_file <- read_csv("/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/HorvathMammalChip_SpeciesGenomeCoordsUnique_v1.0.csv", col_types = cols(.default = "c"))
species_probes <- probe_mapping_file[, c('probeID','MusMusculus')]
species_probes <- as.data.frame(species_probes[!is.na(species_probes$MusMusculus),]) #24048 probes 
species_probes$Chr <- gsub(":.*","",species_probes$MusMusculus)
species_probes$Bp <- species_probes$MusMusculus
species_probes$Bp <- gsub("*.:","", species_probes$Bp)
species_probes <- species_probes[,-2]
species_probes$Position <- paste("chr", paste(species_probes$Chr, species_probes$Bp, sep = ":"), sep ="")
Coordinates <- species_probes

```

All samples are from Cortex Entorhinalis region. There are 40 samples, 20 in each genotype where each 5 samples are taken from ages 6 months, 9 months,10 months and 12 months. 

Beta regression is used as to find sites with genotype differences. Chip is used as a covariate as some samples on the same chip clustered together on heatmaps. 
The p values are FDR adjusted to test for significant sites across the genome. Therefore the threshold on manhattan plots are drawn at -log10 0.05. A table of results is given for sites that pass this significant threshold.


## Single models

### Genotype


$${Methylation } = {Genotype + Age + Genotype*Age + Chip}$$


```{r}
# betaResultsChip <- BetaRegressionArray(betas = betas,
#                            pheno = pheno,
#                            formula = ~Genotype + Age_months+ Genotype*Age_months + Chip_ID,
#                            formulaNull = ~Genotype + Age_months+  Chip_ID,
#                            link = "probit")
# 
# save(betaResultsChip, file = "/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/ArrayBetaRegResults_J20.RData")
```


```{r, warning=F, message=F}

load("/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/ArrayBetaRegResults_J20.RData")
betaResults <- betaResultsChip
# Add chr and bp before deleting rows
Chr <- species_probes$Chr[match(species_probes$probeID, betaResults[,"cpg"])]
Bp <- species_probes$Bp[match(species_probes$probeID, betaResults[,"cpg"])]

betaResults <- cbind(Bp, betaResults)
betaResults <- cbind(Chr, betaResults)

# Remove NA rows
# na_df <- as.data.frame(betaResults[rowSums(is.na(betaResults)) > 0,]) #54 probes has no data in columns
# betaResults <- betaResults[-which(betaResults[,"cpg"] %in% na_df$cpg),] #remove these NAs rows
betaResults <- betaResults[,c(17,1:16)]

# Add FDR pvalues

FDR_adj_genotype <- p.adjust(betaResults[,"p.val.Genotype"], method = "fdr")
FDR_adj_age <- p.adjust(betaResults[,"p.val.Age"], method = "fdr")
FDR_adj_Interaction <- p.adjust(betaResults[,"p.val.Interaction"], method = "fdr")

betaResults <- cbind(betaResults, FDR_adj_genotype, FDR_adj_age, FDR_adj_Interaction)
betaResults <- betaResults[,c(1:10,18,11:13,19,14:17,20)]

# Convert to data frame w/out factors
betaResults2 <- as.data.frame(betaResults)
betaResults2[] <- lapply(betaResults2, as.character)

for(i in 4:20){
  betaResults2[,i] <- as.numeric(as.character(betaResults2[,i]))
}
betaResults2[,"Bp"] <- as.numeric(as.character(betaResults2[,"Bp"]))

betaResults <- betaResults2
# Convert x and y to numbers and remove non autosomal

betaResults<- betaResults[-which(betaResults$Chr %in% c("CHR_MG51_PATCH",
                                                      "CHR_MG4200_PATCH","CHR_MG3699_PATCH")),]
betaResults$Chr <- gsub("X", 20, betaResults$Chr)
betaResults$Chr <- gsub("Y", 21, betaResults$Chr)

betaResults$Chr <- as.numeric(betaResults$Chr)
betaResults$Bp <- as.numeric(betaResults$Bp)

### Add gene names in
mm10_Manifest <- read.csv("/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/mm10_Manifest.csv", stringsAsFactors = F, header = T, row.names = 1)

betaResults$Gene_Symbol <- mm10_Manifest$Gene_Symbol[match(betaResults$cpg, mm10_Manifest$cpg)]
```


```{r}
# Save to outputs

# write.csv(betaResults, file = "/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/DMPs_stats_table_J20.csv")
# 
# 
# sig_genotype <- betaResults[which(betaResults[,"FDR_adj_genotype"] < 0.05),]
# write.csv(sig_genotype, file = "/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/DMPs_J20_sig_genotype.csv")
# 
# sig_age <- betaResults[which(betaResults[,"FDR_adj_age"] < 0.05),]
# write.csv(sig_age, file = "/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/DMPs_J20_sig_age.csv")
# 
# sig_interaction <- betaResults[which(betaResults[,"FDR_adj_Interaction"] < 0.05),]
# write.csv(sig_interaction, file = "/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/DMPs_J20_sig_interaction.csv")


```


```{r, warning=F, message=F}
threshold <- 0.05
betaResults$SNP <- betaResults$Gene_Symbol
color_J20_TG <- "#FF5A62"
rownames(betaResults) <- betaResults$cpg

# Genotype
manhattan(betaResults, p = "FDR_adj_genotype", bp = "Bp", chr = "Chr", 
          genomewide = -log10(threshold), suggestiveline = FALSE, 
          logp=T, col = c("black", color_J20_TG), main = "Genotype",
          chrlabs = c(1:19, "X", "Y"),las = 2, annotatePval = threshold, 
          cex.axis = 1.5, cex.lab = 1.2)

# chromosome problem?
# take chr 10 - 19
# 
# betaResults2 <- betaResults[which(betaResults$Chr == c(10:19)),]
# betaResults2 <- betaResults2[,c(2,3,11)]
# colnames(betaResults2) <- c("CHR", "BP", "P")
# betaResults2$BP = as.integer(format(betaResults2$BP, scientific=F))
# betaResults2$BP <- as.numeric(betaResults2$BP)
# 
# manhattan(betaResults, 
#           genomewide = -log10(threshold), suggestiveline = FALSE, 
#           logp=T)
# 
# manhattan(betaResults2, p = "FDR_adj_genotype", bp = "Bp", chr = "Chr", 
#           genomewide = -log10(threshold), suggestiveline = FALSE, 
#           logp=T, col = c("black", color_J20_TG), main = "Genotype",
#           las = 2, annotatePval = threshold, 
#           cex.axis = 1.5, cex.lab = 1.2)
# 
# for(i in 1:21) {
#   x = sum(is.na(betaResults[which(betaResults$Chr == i),]))
#   print(paste(i, x , sep = " - na:"))
# }
# 
# 
# library(gap)
# ops <- mht.control(colors=rep(c("lightgray","lightblue"),21),srt=0,yline=2.5,xline=2)
# mhtplot(betaResults[,c(2,3,4)])
# mhtplot(betaResults[,c(2,3,11)],ops,xlab="",ylab="",srt=0)
# axis(1,at=1:21)
# axis(2,at=1:16)
```


```{r, warning=F, message=F}
lamda <- qchisq(1-median(betaResults$p.val.Genotype),1)/qchisq(0.5,1)
qq(betaResults$p.val.Genotype, main = "Genotype (Raw pvalue)")
mtext(paste("Lamda = ", signif(lamda,3)), side = 3, adj = 1)

lamda <- qchisq(1-median(betaResults$FDR_adj_genotype),1)/qchisq(0.5,1)
qq(betaResults$FDR_adj_genotype, main = "Genotype (FDR corrected pvalue)")
mtext(paste("Lamda = ", signif(lamda,3)), side = 3, adj = 1)
```

```{r, warning=F, message=F, fig.height = 6, fig.width = 12}
betaResults <- betaResults[order(betaResults$FDR_adj_genotype),]
pheno$Genotype <- factor(pheno$Genotype, levels = c("WT", "TG"))

## Genotypes = with covariates
par(mfrow = c(2,3)) ## will plot 6 figures per page in 2 rows by 3 columns
for(i in 1:6) {
  # model<-lm(as.numeric(betas[rownames(betaResults)[i],]*100) ~ pheno$AgeMonths + pheno$Chip_ID) ## This needs to include all the confouders from your analysis model but not the vaiable you are interested in.
  # adjValues<-residuals(model)+coefficients(model)[1]
  # boxplot(adjValues ~ pheno$Genotype, xlab = "Genotype", ylab = "DNA methylation (%)", 
  #         main = paste(rownames(beta)[i], betaResults[i,"Gene_Symbol"], sep = " ")) ## we multiple by 100 to plot of a % scale rather than proportion
  # mtext(paste("P = ", signif(betaResults[i, "FDR_adj_genotype"],3)), side = 3, adj = 1, cex = 0.7) ## add P value to plot rounded to 3 sf
    plotDMPGenotype(cpg = rownames(betaResults[i,]) , color = color_J20_TG, betaResults = betaResults,
                   betas = betas, pheno = pheno)
  
}
pdf("/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/J20_Genotype.pdf", onefile = T)
for(i in 1:6){
  plotDMPGenotype(cpg = rownames(betaResults[i,]) , color = color_J20_TG, betaResults = betaResults,
                   betas = betas, pheno = pheno)
}
dev.off()
```

```{r, fig.height = 10, fig.width = 50}
if(length(which(betaResults$FDR_adj_genotype <  threshold)) > 0){
  betaResults2 <- betaResults[(which(betaResults$FDR_adj_genotype <  threshold)),]
  betaResults2 <-  betaResults2[order(betaResults2$FDR_adj_genotype),]
  betaResults2 <- betaResults2[,c("cpg","Gene_Symbol", "Chr","Bp","FDR_adj_genotype", "p.val.Genotype","meth.group1.WT",
                             "meth.group2.TG","meth.diff.Genotype" ,"estimate.Genotype","std.error.Genotype")]
  rownames(betaResults2) <- NULL
  colnames(betaResults2) <- c("cpg","Gene Symbol", "Chr","Bp","FDR adj.p.val", "raw p.val","meth WT",
                             "meth TG","meth diff" ,"estimate","std error")
  betaResults2[,c(5,6)] <- format(betaResults2[,c(5,6)], scientific=T)
  kable(betaResults2) %>%
    kable_styling() %>%
    scroll_box(width = "100%", height = "520px")
}
```

### Age

```{r, warning=F, message=F,fig.height = 10, fig.width = 15}
#Age 
manhattan(betaResults, p = "FDR_adj_age", bp = "Bp", chr = "Chr", 
          genomewide = -log10(threshold), suggestiveline = FALSE, 
          logp=T, col = c("black", color_J20_TG), main = "Age",
          chrlabs = c(1:19, "X", "Y"),las = 2, annotatePval = threshold, 
          cex.axis = 1.5, cex.lab = 1.2)

```


```{r, warning=F, message=F}
lamda <- qchisq(1-median(betaResults$p.val.Age),1)/qchisq(0.5,1)
qq(betaResults$p.val.Age, main = "Age (Raw pvalue)")
mtext(paste("Lamda = ", signif(lamda,3)), side = 3, adj = 1)

lamda <- qchisq(1-median(betaResults$FDR_adj_age),1)/qchisq(0.5,1)
qq(betaResults$FDR_adj_age, main = "Age (FDR corrected pvalue)")
mtext(paste("Lamda = ", signif(lamda,3)), side = 3, adj = 1)
```

### Interaction

The individual plots fot sites are the raw methylation values of the samples across ages. The dotted lines are the mean values for across the ages within genotypes. 


```{r, warning=F, message=F, fig.height = 10, fig.width = 15}
#Interaction
manhattan(betaResults, p = "FDR_adj_Interaction", bp = "Bp", chr = "Chr", 
          genomewide = -log10(threshold), suggestiveline = FALSE, 
          logp=T, col = c("black", color_J20_TG), main = "Genotype*Age Interaction",
          chrlabs = c(1:19, "X", "Y"),las = 2, 
          cex.axis = 1.5, cex.lab = 1.2)

```


```{r, warning=F, message=F}
lamda <- qchisq(1-median(betaResults$p.val.Interaction),1)/qchisq(0.5,1)
qq(betaResults$p.val.Interaction, main = "Genotype*Age Interaction (Raw pvalue)")
mtext(paste("Lamda = ", signif(lamda,3)), side = 3, adj = 1)

lamda <- qchisq(1-median(betaResults$FDR_adj_Interaction),1)/qchisq(0.5,1)
qq(betaResults$FDR_adj_Interaction, main = "Interaction (FDR corrected pvalue)")
mtext(paste("Lamda = ", signif(lamda,3)), side = 3, adj = 1)
```


```{r, warning=F, message=F, fig.height = 6, fig.width = 12}
betaResults <- betaResults[order(betaResults$FDR_adj_Interaction),]
pheno$Genotype <- factor(pheno$Genotype, levels = c("WT", "TG"))
pheno$col <-  ifelse(pheno$Genotype == "WT","#000000", color_J20_TG)

SamplesMean <- function(Genotype = Genotype, Age = Age, betas = betas, pheno = pheno, cpg = cpg){
    samples <- rownames(pheno[which(pheno$Genotype == Genotype & pheno$AgeMonths == Age),])
    betas_samples <-  betas[cpg,samples]
    mean_samples <- apply(betas_samples,1, mean)

    return(mean_samples)
  }

geno_age_mean <- matrix(NA, nrow = 4, ncol = 2)
colnames(geno_age_mean) <- c("WT", "TG")
rownames(geno_age_mean) <- c(6,8,10,12)
ages <- c(6,8,10,12)

par(mfrow = c(2,3)) ## will plot 6 figures per page in 2 rows by 3 columns
for(i in 1:6){
  # for( s in 1:4){
  #   j = ages[s]
  #   geno_age_mean[s,1] <- SamplesMean("WT", j, betas, pheno, cpg = rownames(betaResults)[i])
  #   geno_age_mean[s,2] <- SamplesMean("TG", j, betas, pheno, cpg = rownames(betaResults)[i])
  # }
  # geno_age_mean <- as.data.frame(geno_age_mean)
  # 
  # plot(pheno$AgeMonths, as.numeric(betas[rownames(betaResults)[i],]*100),
  #      xlab = "Age (months)", ylab = "DNA methylation (%)", pch = 16,
  #      main = paste(rownames(betas)[i], betaResults[i,"Gene_Symbol"], sep = " "),
  #      col = pheno$col)
  # lines(as.numeric(rownames(geno_age_mean)), geno_age_mean$WT*100, col = "#000000", lty = 2)
  # lines(as.numeric(rownames(geno_age_mean)), geno_age_mean$TG*100, col = color_J20_TG, lty = 2)
  # 
  # par(xpd  = TRUE)
  # legend(x ="topright",legend = levels(factor(pheno$Genotype)),
  #        col = c("#000000", color_J20_TG), pch = 19, cex = .7,
  #        inset=c(-0.02,-0.25))
  # 
  # par(xpd  = FALSE)
  # mtext(paste("P = ", signif(betaResults[i, "FDR_adj_Interaction"],3)), side = 3, adj = 1, cex = 0.7) ## add P value to plot rounded to 3 sf
  
    plotDMPInteraction(cpg = rownames(betaResults[i,]) , color = color_J20_TG, betaResults = betaResults,
                   betas = betas, pheno = pheno)
  
}
pdf("/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/J20_Interaction.pdf", onefile = T)
for(i in 1:6){
  plotDMPInteraction(cpg = rownames(betaResults[i,]) , color = color_J20_TG, betaResults = betaResults,
                   betas = betas, pheno = pheno)
}
dev.off()

```

### Pathology


$${Methylation } = {Pathology + Chip}$$

```{r}
# betaResultsPathology <- BetaRegressionArrayPathology(betas = betas,
#                                                      pheno = pheno,
#                                                      formula = ~Pathology_ECX + Chip_ID,
#                                                      link = "probit")
# resave(betaResultsPathology, file = "/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/ArrayBetaRegResults_J20.RData")
```

```{r, warning=F, message=F}

betaResults <- betaResultsPathology
# Add chr and bp before deleting rows
Chr <- species_probes$Chr[match(species_probes$probeID, betaResults[,"cpg"])]
Bp <- species_probes$Bp[match(species_probes$probeID, betaResults[,"cpg"])]

betaResults <- cbind(Bp, betaResults)
betaResults <- cbind(Chr, betaResults)

# Remove NA rows
# na_df <- as.data.frame(betaResults[rowSums(is.na(betaResults)) > 0,]) #54 probes has no data in columns
# betaResults <- betaResults[-which(betaResults[,"cpg"] %in% na_df$cpg),] #remove these NAs rows
betaResults <- betaResults[,c(10,1:9)]

# Add FDR pvalues

FDR_adj_pathology <- p.adjust(betaResults[,"p.val.Pathology"], method = "fdr")

betaResults <- cbind(betaResults, FDR_adj_pathology)
betaResults <- betaResults[,c(1:4,11,5:10)]

# Convert to data frame w/out factors
betaResults2 <- as.data.frame(betaResults)
betaResults2[] <- lapply(betaResults2, as.character)

for(i in 4:11){
  betaResults2[,i] <- as.numeric(as.character(betaResults2[,i]))
}
betaResults2[,"Bp"] <- as.numeric(as.character(betaResults2[,"Bp"]))
betaResults2[,"cpg"] <- as.character(betaResults2[,"cpg"]) 
betaResults <- betaResults2

# Convert x and y to numbers and remove non autosomal

betaResults<- betaResults[-which(betaResults$Chr %in% c("CHR_MG51_PATCH",
                                                      "CHR_MG4200_PATCH","CHR_MG3699_PATCH")),]
betaResults$Chr <- gsub("X", 20, betaResults$Chr)
betaResults$Chr <- gsub("Y", 21, betaResults$Chr)

betaResults$Chr <- as.numeric(betaResults$Chr)
betaResults$Bp <- as.numeric(betaResults$Bp)

### Add gene names in
mm10_Manifest <- read.csv("/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/mm10_Manifest.csv", stringsAsFactors = F, header = T, row.names = 1)

betaResults$Gene_Symbol <- mm10_Manifest$Gene_Symbol[match(betaResults$cpg, mm10_Manifest$cpg)]
``` 



```{r, warning=F, message=F,fig.height = 10, fig.width = 15}
threshold <- 0.05
betaResults$SNP <- betaResults$Gene_Symbol
color_J20_TG <- "#FF5A62"


# Pathology
manhattan(betaResults, p = "FDR_adj_pathology", bp = "Bp", chr = "Chr", 
          genomewide = -log10(threshold), suggestiveline = FALSE, 
          logp=T, col = c("black", color_J20_TG), main = "Pathology",
          chrlabs = c(1:19, "X", "Y"),las = 2, annotatePval = threshold,
          cex.axis = 1.5, cex.lab = 1.2)


```


```{r, warning=F, message=F}
lamda <- qchisq(1-median(betaResults$p.val.Pathology),1)/qchisq(0.5,1)
qq(betaResults$p.val.Pathology, main = "Pathology (Raw pvalue")
mtext(paste("Lamda = ", signif(lamda,3)), side = 3, adj = 1)

lamda <- qchisq(1-median(betaResults$FDR_adj_pathology),1)/qchisq(0.5,1)
qq(betaResults$FDR_adj_pathology, main = "Pathology (FDR corrected pvalue)")
mtext(paste("Lamda = ", signif(lamda,3)), side = 3, adj = 1)
```

```{r, warning=F, message=F, fig.height = 6, fig.width = 12}
betaResults <- betaResults[order(betaResults$FDR_adj_pathology),]
rownames(betaResults) <- betaResults$cpg

pheno$col <-  ifelse(pheno$Genotype == "WT","#000000", color_J20_TG )

## Pathologys
par(mfrow = c(2,3)) ## will plot 6 figures per page in 2 rows by 3 columns
for(i in 1:6){
  plot(pheno$Pathology_ECX, as.numeric(betas[rownames(betaResults)[i],]*100),
       xlab = "Pathology", ylab = "DNA methylation (%)", pch = 16,
       main = paste(rownames(betas)[i], betaResults[i,"Gene_Symbol"], sep = " "),
       col = pheno$col) ## we multiple by 100 to plot of a % scale rather than proportion
  ## to add line of best fit
  mtext(paste("P = ", signif(betaResults[i, "FDR_adj_pathology"],3)), side = 3, adj = 0, cex = .7) ## add P value to plot rounded to 3 sf
   par(xpd  = TRUE)
  legend(x ="topright",legend = levels(factor(pheno$Genotype)),
         col = c("#000000",color_J20_TG), pch = 19, cex = .7,
         inset=c(-0.02,-0.25))
}

for(i in 1:6){
  plotDMPPathology(cpg = rownames(betaResults[i,]) , color = color_J20_TG, betaResults = betaResults,
                   betas = betas, pheno = pheno)
}


pdf("/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/J20_Pathology.pdf", onefile = T)
for(i in 1:6){
  plotDMPPathology(cpg = rownames(betaResults[i,]) , color = color_J20_TG, betaResults = betaResults,
                   betas = betas, pheno = pheno)
}
dev.off()
```

```{r, warning=F, message=F,fig.height = 10, fig.width = 50}
if(length(which(betaResults$FDR_adj_pathology <  threshold)) > 0){
  betaResults2 <- betaResults[(which(betaResults$FDR_adj_pathology <  threshold)),]
  betaResults2 <-  betaResults2[order(betaResults2$FDR_adj_pathology),]
  betaResults2 <- betaResults2[,c("cpg","Gene_Symbol", "Chr","Bp","FDR_adj_pathology", "p.val.Pathology",
                                  "estimate.Pathology","std.error.Pathology")]
  rownames(betaResults2) <- NULL
  colnames(betaResults2) <- c("cpg","Gene Symbol", "Chr","Bp","FDR adj p.val", "raw p.val",
                                  "estimate","std error")
  betaResults2[,c(5,6)] <- format(betaResults2[,c(5,6)], scientific=T)
  kable(betaResults2) %>%
    kable_styling() %>%
    scroll_box(width = "100%", height = "520px")
}

```

```{r}
#save outputs

# write.csv(betaResults, file = "/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/DMPs_stats_table_J20_Pathology.csv")
# 
# 
# sig_pathology <- betaResults[which(betaResults[,"FDR_adj_pathology"] < 0.05),]
# write.csv(sig_pathology, file = "/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/DMPs_J20_sig_pathology.csv")

```

## Full model

 $${Methylation } = {Genotype + Age + Pathology + Chip}$$

```{r, warning=F, message = F}

# betaResultsChip <- BetaRegressionFullModel(betas = betas,
#                            pheno = pheno,
#                            formula = ~Genotype + Age_months+ Pathology_ECX + Chip_ID,
#                            link = "probit")
# 
# save(betaResultsChip, file = "/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/ArrayBetaRegResultsFull_J20.RData")

```



```{r, warning=F, message=F}

load("/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/ArrayBetaRegResultsFull_J20.RData")
betaResults <- betaResultsChip
# Add chr and bp before deleting rows
Chr <- species_probes$Chr[match(species_probes$probeID, betaResults[,"cpg"])]
Bp <- species_probes$Bp[match(species_probes$probeID, betaResults[,"cpg"])]

betaResults <- cbind(Bp, betaResults)
betaResults <- cbind(Chr, betaResults)



# Add FDR pvalues

FDR_adj_genotype <- p.adjust(betaResults[,"p.val.Genotype"], method = "fdr")
FDR_adj_age <- p.adjust(betaResults[,"p.val.Age"], method = "fdr")
FDR_adj_pathology <- p.adjust(betaResults[,"p.val.Pathology"], method = "fdr")

betaResults <- cbind(betaResults, FDR_adj_genotype, FDR_adj_age, FDR_adj_pathology)
betaResults <- betaResults[,c(16,1:15,17:19)]
# Convert to data frame w/out factors
betaResults2 <- as.data.frame(betaResults)
betaResults2[] <- lapply(betaResults2, as.character)

for(i in 3:ncol(betaResults2)){
  betaResults2[,i] <- as.numeric(as.character(betaResults2[,i]))
}
betaResults2[,"Bp"] <- as.numeric(as.character(betaResults2[,"Bp"]))

betaResults <- betaResults2
# Convert x and y to numbers and remove non autosomal

betaResults<- betaResults[-which(betaResults$Chr %in% c("CHR_MG51_PATCH",
                                                      "CHR_MG4200_PATCH","CHR_MG3699_PATCH")),]
betaResults$Chr <- gsub("X", 20, betaResults$Chr)
betaResults$Chr <- gsub("Y", 21, betaResults$Chr)

betaResults$Chr <- as.numeric(betaResults$Chr)
betaResults$Bp <- as.numeric(betaResults$Bp)

### Add gene names in
mm10_Manifest <- read.csv("/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/mm10_Manifest.csv", stringsAsFactors = F, header = T, row.names = 1)

betaResults$Gene_Symbol <- mm10_Manifest$Gene_Symbol[match(betaResults$cpg, mm10_Manifest$cpg)]
```

### Genotype

```{r, warning=F, message=F}
threshold <- 0.05
betaResults$SNP <- betaResults$Gene_Symbol
color_J20_TG <- "#FF5A62"
rownames(betaResults) <- betaResults$cpg

# Genotype
manhattan(betaResults, p = "FDR_adj_genotype", bp = "Bp", chr = "Chr", 
          genomewide = -log10(threshold), suggestiveline = FALSE, 
          logp=T, col = c("black", color_J20_TG), main = "Genotype",
          chrlabs = c(1:19, "X", "Y"),las = 2, annotatePval = threshold, 
          cex.axis = 1.5, cex.lab = 1.2)


```


```{r, warning=F, message=F}
lamda <- qchisq(1-median(betaResults$p.val.Genotype),1)/qchisq(0.5,1)
qq(betaResults$p.val.Genotype, main = "Genotype (Raw pvalue)")
mtext(paste("Lamda = ", signif(lamda,3)), side = 3, adj = 1)

lamda <- qchisq(1-median(betaResults$FDR_adj_genotype),1)/qchisq(0.5,1)
qq(betaResults$FDR_adj_genotype, main = "Genotype (FDR corrected pvalue)")
mtext(paste("Lamda = ", signif(lamda,3)), side = 3, adj = 1)
```

```{r, warning=F, message=F}

betaResults <- betaResults[order(betaResults$FDR_adj_genotype),]
pheno$Genotype <- factor(pheno$Genotype, levels = c("WT", "TG"))

## Genotypes = with covariates
par(mfrow = c(2,3)) ## will plot 6 figures per page in 2 rows by 3 columns
for(i in 1:6) {
  # model<-lm(as.numeric(betas[rownames(betaResults)[i],]*100) ~ pheno$AgeMonths + pheno$Chip_ID) ## This needs to include all the confouders from your analysis model but not the vaiable you are interested in.
  # adjValues<-residuals(model)+coefficients(model)[1]
  # boxplot(adjValues ~ pheno$Genotype, xlab = "Genotype", ylab = "DNA methylation (%)", 
  #         main = paste(rownames(beta)[i], betaResults[i,"Gene_Symbol"], sep = " ")) ## we multiple by 100 to plot of a % scale rather than proportion
  # mtext(paste("P = ", signif(betaResults[i, "FDR_adj_genotype"],3)), side = 3, adj = 1, cex = 0.7) ## add P value to plot rounded to 3 sf
    plotDMPGenotype(cpg = rownames(betaResults[i,]) , color = color_J20_TG, betaResults = betaResults,
                   betas = betas, pheno = pheno)
  
}
pdf("/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/FullJ20_Genotype.pdf", onefile = T)
for(i in 1:6){
  plotDMPGenotype(cpg = rownames(betaResults[i,]) , color = color_J20_TG, betaResults = betaResults,
                   betas = betas, pheno = pheno)
}
dev.off()
```

```{r, fig.height = 10, fig.width = 50}
if(length(which(betaResults$FDR_adj_genotype <  threshold)) > 0){
  betaResults2 <- betaResults[(which(betaResults$FDR_adj_genotype <  threshold)),]
  betaResults2 <-  betaResults2[order(betaResults2$FDR_adj_genotype),]
  betaResults2 <- betaResults2[,c("cpg","Gene_Symbol", "Chr","Bp","FDR_adj_genotype", "p.val.Genotype","meth.group1.WT",
                             "meth.group2.TG","meth.diff.Genotype" ,"estimate.Genotype","std.error.Genotype")]
  rownames(betaResults2) <- NULL
  colnames(betaResults2) <- c("cpg","Gene Symbol", "Chr","Bp","FDR adj.p.val", "raw p.val","meth WT",
                             "meth TG","meth diff" ,"estimate","std error")
  betaResults2[,c(5,6)] <- format(betaResults2[,c(5,6)], scientific=T)
  kable(betaResults2) %>%
    kable_styling() %>%
    scroll_box(width = "100%", height = "520px")
}
```



### Pathology

```{r, warning=F, message=F,fig.height = 10, fig.width = 15}
threshold <- 0.05
betaResults$SNP <- betaResults$Gene_Symbol
color_J20_TG <- "#FF5A62"


# Pathology
manhattan(betaResults, p = "FDR_adj_pathology", bp = "Bp", chr = "Chr", 
          genomewide = -log10(threshold), suggestiveline = FALSE, 
          logp=T, col = c("black", color_J20_TG), main = "Pathology",
          chrlabs = c(1:19, "X", "Y"),las = 2, annotateTop = threshold,
          cex.axis = 1.5, cex.lab = 1.2)


```


```{r, warning=F, message=F}
lamda <- qchisq(1-median(betaResults$p.val.Pathology),1)/qchisq(0.5,1)
qq(betaResults$p.val.Pathology, main = "Pathology (Raw pvalue)")
mtext(paste("Lamda = ", signif(lamda,3)), side = 3, adj = 1)

lamda <- qchisq(1-median(betaResults$FDR_adj_pathology),1)/qchisq(0.5,1)
qq(betaResults$FDR_adj_pathology, main = "Pathology (FDR corrected pvalue)")
mtext(paste("Lamda = ", signif(lamda,3)), side = 3, adj = 1)
```


```{r, warning=F, message=F, fig.height = 6, fig.width = 12}
betaResults <- betaResults[order(betaResults$FDR_adj_pathology),]
rownames(betaResults) <- betaResults$cpg

pheno$col <-  ifelse(pheno$Genotype == "WT","#000000", color_J20_TG )

## Pathologys
par(mfrow = c(2,3)) ## will plot 6 figures per page in 2 rows by 3 columns
for(i in 1:6){
  plot(pheno$Pathology_HIP, as.numeric(betas[rownames(betaResults)[i],]*100),
       xlab = "Pathology", ylab = "DNA methylation (%)", pch = 16,
       main = paste(rownames(betas)[i], betaResults[i,"Gene_Symbol"], sep = " "),
       col = pheno$col) ## we multiple by 100 to plot of a % scale rather than proportion
  ## to add line of best fit
  mtext(paste("P = ", signif(betaResults[i, "FDR_adj_pathology"],3)), side = 3, adj = 0, cex = .7) ## add P value to plot rounded to 3 sf
   par(xpd  = TRUE)
  legend(x ="topright",legend = levels(factor(pheno$Genotype)),
         col = c("#000000",color_J20_TG), pch = 19, cex = .7,
         inset=c(-0.02,-0.25))
}

for(i in 1:6){
  plotDMPPathology(cpg = rownames(betaResults[i,]) , color = color_J20_TG, betaResults = betaResults,
                   betas = betas, pheno = pheno)
}

pdf("/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/FullJ20_Pathology.pdf", onefile = T)
for(i in 1:6){
  plotDMPPathology(cpg = rownames(betaResults[i,]) , color = color_J20_TG, betaResults = betaResults,
                   betas = betas, pheno = pheno)
}
dev.off()
```

```{r, warning=F, message=F,fig.height = 10, fig.width = 50}
if(length(which(betaResults$FDR_adj_pathology <  threshold)) > 0){
  betaResults2 <- betaResults[(which(betaResults$FDR_adj_pathology <  threshold)),]
  betaResults2 <-  betaResults2[order(betaResults2$FDR_adj_pathology),]
  betaResults2 <- betaResults2[,c("cpg","Gene_Symbol", "Chr","Bp","FDR_adj_pathology", "p.val.Pathology",
                                  "estimate.Pathology","std.error.Pathology")]
  rownames(betaResults2) <- NULL
  colnames(betaResults2) <- c("cpg","Gene Symbol", "Chr","Bp","FDR adj p.val", "raw p.val",
                                  "estimate","std error")
  betaResults2[,c(5,6)] <- format(betaResults2[,c(5,6)], scientific=T)
  kable(betaResults2) %>%
    kable_styling() %>%
    scroll_box(width = "100%", height = "520px")
}

```




```{r}
# Save to outputs

write.csv(betaResults, file = "/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/DMPs_stats_table_FullJ20.csv")


sig_genotype <- betaResults[which(betaResults[,"FDR_adj_genotype"] < 0.05),]
write.csv(sig_genotype, file = "/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/DMPs_FullJ20_sig_genotype.csv")

sig_age <- betaResults[which(betaResults[,"FDR_adj_age"] < 0.05),]
write.csv(sig_age, file = "/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/DMPs_FullJ20_sig_age.csv")

sig_pathology <- betaResults[which(betaResults[,"FDR_adj_pathology"] < 0.05),]
write.csv(sig_pathology, file = "/gpfs/ts0/projects/Research_Project-191406/Aisha/data/Array/DMPs_FullJ20_sig_interaction.csv")


```



